protocol-tokio/src/protocol/codec/message.rs

```diff
@@ -88,17 +89,113 @@ pub enum Message {
     ProbeSocketAck(nt::LightWeightConnectionId),
     CreateNodeId(nt::LightWeightConnectionId, NodeId),
     AckNodeId(nt::LightWeightConnectionId, NodeId),
+    Bytes(nt::LightWeightConnectionId, Bytes),
+
+    UserMessage(M),
+}
 
-    GetBlockHeaders(nt::LightWeightConnectionId, GetBlockHeaders),
-    BlockHeaders(nt::LightWeightConnectionId, Response<BlockHeaders, String>),
-    GetBlocks(nt::LightWeightConnectionId, GetBlocks),
-    Block(nt::LightWeightConnectionId, Response<Block, String>),
-    SendTransaction(nt::LightWeightConnectionId, TxAux),
+pub enum ChainMessage<B, T>
+where
+    B: server::block::BlockService,
+    T: server::transaction::TransactionService,
+{
+    GetBlockHeaders(nt::LightWeightConnectionId, GetBlockHeaders<B::BlockId>),
+    BlockHeaders(
+        nt::LightWeightConnectionId,
+        Response<BlockHeaders<B::Header>, String>,
+    ),
+    GetBlocks(nt::LightWeightConnectionId, GetBlocks<B::BlockId>),
+    Block(nt::LightWeightConnectionId, Response<B::Block, String>),
+    SendTransaction(nt::LightWeightConnectionId, T::TransactionId),
     TransactionReceived(nt::LightWeightConnectionId, Response<bool, String>),
     Subscribe(nt::LightWeightConnectionId, KeepAlive),
-    Bytes(nt::LightWeightConnectionId, Bytes),
 }
-impl Message {
+
+pub trait AsNtEvent
+where
+    Self: std::marker::Sized,
+{
+    fn to_nt(self) -> nt::Event;
+    fn try_from_nt(lwcid: nt::LightWeightConnectionId, bytes: &Bytes) -> Option<Self>;
+}
+
+pub struct Any();
```

Comment:
What is this for?

Reply:
This is a hack, any better idea are welcome, the problem is that for the type inference rustc should know the type of the application specific payload, but `accepting` and `connecting` traits do not care about it, as they do not exchange the application specific payload. `Any` introduces a special type that we can use there so those structures work for any user payload.